[REBASE & FF] Q35/SBSA: Enable Debugger By Default in Debug Builds

os-d opened the pull request:
## Description

In order to show a realistic platform example, this commit enables the debugger by default in debug builds. This ensures that the file space required for it will always be available in debug builds. It is disabled by default in release builds to save the space for production.

This also fixes:

 Makefile.toml: Unify Make Arg Handling
--
Commit e3dd8615044ad2e9dafe9801840fece4cf40f2e9 was tested using patina-qemu's build and run script, which passes arguments in a different fashion than directly running cargo make. As a result, using that script worked with the Makefile.toml changes introduced in the above commit, but running the same commands locally could fail to produce the desired result. This commit ensures that the cargo make syntax is respected and a separate PR is fixing patina-qemu's build and run script so it passed arguments in the same order as cargo make naturally does.

- [ ] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

Tested building and booting.

## Integration Instructions

N/A.

os-d commented on Makefile.toml:
```diff
@@ -287,35 +287,25 @@ set_env COMBINED_FEATURES ${combined_features}
 description = """Builds the DEBUG Q35 UEFI firmware."""
-env = { CARGO_BIN_NAME = "qemu_q35_dxe_core", "RUSTC_PROFILE" = "dev" }
-run_task = "build-efi-x64"
+env = { CARGO_BIN_NAME = "qemu_q35_dxe_core", "RUSTC_PROFILE" = "dev", CARGO_BIN_TARGET = "x86_64-unknown-uefi", CARGO_BIN_FEATURES = "x64,enable_debugger" }
```
Need to fix this, the last one enabled the initial breakpoint always, will fix.

os-d replied:
fixed.

Javagedes commented:
I would prefer you don't remove the "tier"ing that I implemented, but I won't block it like this.

A way you can keep it tiered, is to update the `build-efi-x64` and `build-efi-aarch64` to be

`env = { CARGO_BIN_TARGET = "x86_64-unknown-uefi", CARGO_BIN_FEATURES = "x64,${CARGO_BIN_FEATURES}" }`
`env = { CARGO_BIN_TARGET = "aarch64-unknown-uefi", CARGO_BIN_FEATURES = "aarch64,${CARGO_BIN_FEATURES}" }`

and add `CARGO_BIN_FEATURES = { value = "", condition = { env_not_set = ["CARGO_BIN_FEATURES"] } }` at the top level [env].

Then update the q35 / sbsa commands to be

```
[tasks.q35]
description = """Builds the DEBUG Q35 UEFI firmware."""
env = { CARGO_BIN_NAME = "qemu_q35_dxe_core", "RUSTC_PROFILE" = "dev", CARGO_BIN_FEATURES = "enable_debugger"}
run_task = "build-efi-x64"
```

This is more extensible down the road if we have multiple togglable features

os-d commented:
> I would prefer you don't remove the "tier"ing that I implemented, but I won't block it like this.
> 
> A way you can keep it tiered, is to update the `build-efi-x64` and `build-efi-aarch64` to be
> 
> `env = { CARGO_BIN_TARGET = "x86_64-unknown-uefi", CARGO_BIN_FEATURES = "x64,${CARGO_BIN_FEATURES}" }` `env = { CARGO_BIN_TARGET = "aarch64-unknown-uefi", CARGO_BIN_FEATURES = "aarch64,${CARGO_BIN_FEATURES}" }`
> 
> and add `CARGO_BIN_FEATURES = { value = "", condition = { env_not_set = ["CARGO_BIN_FEATURES"] } }` at the top level [env].
> 
> Then update the q35 / sbsa commands to be
> 
> ```
> [tasks.q35]
> description = """Builds the DEBUG Q35 UEFI firmware."""
> env = { CARGO_BIN_NAME = "qemu_q35_dxe_core", "RUSTC_PROFILE" = "dev", CARGO_BIN_FEATURES = "enable_debugger"}
> run_task = "build-efi-x64"
> ```
> 
> This is more extensible down the road if we have multiple togglable features

I can leave it, I just noticed that with the recent refactoring, those wrapped ones didn't do anything and we have multiple layers of wrapping, which is why I removed this layer. But I'm fine to leave it if you think it adds value.

Javagedes commented:
> > I would prefer you don't remove the "tier"ing that I implemented, but I won't block it like this.
> > A way you can keep it tiered, is to update the `build-efi-x64` and `build-efi-aarch64` to be
> > `env = { CARGO_BIN_TARGET = "x86_64-unknown-uefi", CARGO_BIN_FEATURES = "x64,${CARGO_BIN_FEATURES}" }` `env = { CARGO_BIN_TARGET = "aarch64-unknown-uefi", CARGO_BIN_FEATURES = "aarch64,${CARGO_BIN_FEATURES}" }`
> > and add `CARGO_BIN_FEATURES = { value = "", condition = { env_not_set = ["CARGO_BIN_FEATURES"] } }` at the top level [env].
> > Then update the q35 / sbsa commands to be
> > ```
> > [tasks.q35]
> > description = """Builds the DEBUG Q35 UEFI firmware."""
> > env = { CARGO_BIN_NAME = "qemu_q35_dxe_core", "RUSTC_PROFILE" = "dev", CARGO_BIN_FEATURES = "enable_debugger"}
> > run_task = "build-efi-x64"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is more extensible down the road if we have multiple togglable features
> 
> I can leave it, I just noticed that with the recent refactoring, those wrapped ones didn't do anything and we have multiple layers of wrapping, which is why I removed this layer. But I'm fine to leave it if you think it adds value.

I think it leaves the possibility of adding value if we add more toggle-able features in the future.

cfernald commented on bin/q35_dxe_core.rs:
```diff
@@ -47,10 +47,15 @@ static LOGGER: AdvancedLogger<Uart16550> = AdvancedLogger::new(
 static DEBUGGER: patina_debugger::PatinaDebugger<Uart16550> =
     patina_debugger::PatinaDebugger::new(Uart16550::Io { base: 0x3F8 })
-        .with_force_enable(true)
+        .with_force_enable(_ENABLE_INITIAL_BREAKPOINT)
```
I think this is post-checking, but this is a little misleading as this will control full enablement, not just the initial breakpoint

os-d replied:
Discussing on other PR thread.
